File: ios/Sources/RoadSenseNSBootstrap/Pipeline/RoughnessScorer.swift
import Foundation

public struct RoughnessScorer: Sendable {
    public let sampleRateHz: Double
    public let cutoffHz: Double
    public let settleSampleCount: Int

    public init(
        sampleRateHz: Double = 50,
        cutoffHz: Double = 0.5,
        settleSampleCount: Int? = nil
    ) {
        self.sampleRateHz = sampleRateHz
        self.cutoffHz = cutoffHz
        self.settleSampleCount = settleSampleCount ?? Int(sampleRateHz)
    }

    public func score(samples: [MotionSample]) -> Double {
        score(verticalAccelerations: samples.map(\.verticalAcceleration))
    }

    public func score(verticalAccelerations: [Double]) -> Double {
        guard !verticalAccelerations.isEmpty else {
            return 0
        }

        var filter = HighPassBiquad.makeButterworth(
            sampleRateHz: sampleRateHz,
            cutoffHz: cutoffHz
        )

        if let firstSample = verticalAccelerations.first {
            for _ in 0..<settleSampleCount {
                _ = filter.apply(firstSample)
            }
        }

        let filtered = verticalAccelerations.map { filter.apply($0) }
        let meanSquare = filtered.reduce(0.0) { partialResult, sample in
            partialResult + (sample * sample)
        } / Double(filtered.count)

        return sqrt(meanSquare)
    }
}

File: android/core-sensor/src/main/kotlin/ca/roadsense/ns/sensor/RoughnessScorer.kt
package ca.roadsense.ns.sensor

import kotlin.math.sqrt

class RoughnessScorer(
    val sampleRateHz: Double = 50.0,
    val cutoffHz: Double = 0.5,
    settleSampleCount: Int? = null,
) {
    val settleSampleCount: Int = settleSampleCount ?: sampleRateHz.toInt()

    fun score(samples: List<MotionSample>): Double =
        scoreVerticalAccelerations(samples.map { it.verticalAcceleration })

    fun scoreVerticalAccelerations(verticalAccelerations: List<Double>): Double {
        if (verticalAccelerations.isEmpty()) return 0.0

        val filter = HighPassBiquad.makeButterworth(sampleRateHz, cutoffHz)

        val first = verticalAccelerations.first()
        repeat(settleSampleCount) { filter.apply(first) }

        val filtered = verticalAccelerations.map { filter.apply(it) }
        val meanSquare = filtered.sumOf { it * it } / filtered.size.toDouble()
        return sqrt(meanSquare)
    }
}
